src/macros.rs

```diff
@@ -0,0 +1,239 @@
+            type Item = $crate::Result<$crate::Multipart>;
+
+            fn poll_next(
+                self: ::std::pin::Pin<&mut Self>,
+                cx: &mut ::std::task::Context,
+            ) -> ::std::task::Poll<::std::option::Option<Self::Item>> {
+                match self.$socket.multipart_recv(cx) {
+                    ::std::task::Poll::Ready(value) => {
+                        ::std::task::Poll::Ready(::std::option::Option::Some(value))
+                    }
+                    _ => ::std::task::Poll::Pending,
+                }
+            }
+        }
+    };
+}
+/// Implements a buffered `Stream<Item=Result<Multipart>>` for the given type.
+/// $buffer: identifier of a field containing a `ReceiveBuffer`
+/// $socket: identifier of a field containing a `ZmqPoller`
+macro_rules! impl_buffered_stream {
+    ($type: ty, $buffer: ident, $socket: ident) => {
+        impl ::futures::Stream for $type {
+            type Item = $crate::Result<$crate::Multipart>;
+
+            fn poll_next(
+                self: ::std::pin::Pin<&mut Self>,
+                cx: &mut ::std::task::Context,
+            ) -> ::std::task::Poll<::std::option::Option<Self::Item>> {
+                let Self {
+                    ref $socket,
+                    ref mut $buffer,
+                } = self.get_mut();
+                match $socket.multipart_recv_buffered(cx, $buffer) {
+                    ::std::task::Poll::Ready(value) => {
+                        ::std::task::Poll::Ready(::std::option::Option::Some(value))
+                    }
+                    _ => ::std::task::Poll::Pending,
+                }
+            }
+        }
+    };
+}
+
+/// Builder implementations
+
+/// Creates a builder structure named $builder with methods to bind and connect using the given $socket_type.
+/// The $result_type must have a constructor accepting a single field called `socket` of type [`zmq::Socket`].
+macro_rules! impl_builder {
+    ($socket_type: ident, $builder: ident, $result_type: tt) => {
+        pub struct $builder<'a> {
+            context: &'a ::zmq::Context,
+        }
+
+        impl<'a> $builder<'a> {
+            fn new(context: &'a ::zmq::Context) -> Self {
+                Self { context }
+            }
+
+            pub fn connect(self, endpoint: &str) -> $crate::Result<$result_type> {
+                let socket = self.context.socket(::zmq::SocketType::$socket_type)?;
+                socket.connect(endpoint)?;
+
+                $crate::Result::Ok($result_type {
+                    socket: socket.into(),
+                })
+            }
+            pub fn bind(self, endpoint: &str) -> $crate::Result<$result_type> {
+                let socket = self.context.socket(::zmq::SocketType::$socket_type)?;
+                socket.bind(endpoint)?;
+
+                $crate::Result::Ok($result_type {
+                    socket: socket.into(),
+                })
+            }
+        }
+    };
+}
```

Comment:
Is there any way to write documentation for this struct?  I don't know too much about rust macros, but neither doc comments nor `#[doc="foo"]` worked.  In fact, the output of `cargo doc` doesn't even include the builder structs at all! 😢

Reply:
For a while I thought that it's a bug in `cargo doc`, but it's actually caused by the fact that these structs are not exported. If you export them (for example `pub use request_reply::*;` in `socket_types/mod.rs`), they show up in the documentation.

Reply:
Interesting.  Alternative: since what we're doing is the same for each socket type, we could do it via a shared struct rather than manually stamping out the code with a macro:

e.g. (experimental, keeps API the same, passes tests)
https://github.com/skrap/tmq/commit/9654ba036abe7d0403a02c05ad5cb7c8b88d8b0b

This would let us expose one set of socket builder documentation, rather than one per socket type.

Reply:
Indeed this is a better solution. I only used macros because I wanted to differ between bind/connect, but since both are implemented for all sockets, it's unnecessary.

But the `From` impl should be handled differently I think. Currently it's only possible to create the `*BuilderBound` structs from within the file where they exist (since their `socket` field is private). This upholds the invariant that the passed ZMQ socket is indeed bound or connected. If we exposed a `From` impl, a user could convert an unconnected/unbound ZMQ socket into the `*BuilderBound` struct. It's probably a minor thing, but I think that it shouldn't be that hard to avoid this. For example by introducing a new (crate local) trait to convert the `*BuilderBound` structures from `zmq::Socket`.

(In the meantime I pushed a commit which exposes more things in the documentation. If you think it's better, we can use this style and document the rest of the sockets.)

Reply:
Since the docs on the macro'd version work fine, let's go with that.  I agree on the unease about the public-ness of using the `From` trait.  I retract my suggestion :)

Reply:
Actually "retract" is too strong — I'm really fine going forward with any of the above solutions.  I'll let you pick what you like best.

Reply:
I realized that there was too much unneeded cruft. I removed one layer of builder indirection (it was only really needed for `SUB` sockets) and removed the builder macro (I built on top of your commit). Ironically, the private trait impl (I called it `FromZmqSocket`) doesn't work, since it leaks into a public interface (`SocketBuilder`). However, since I hid it from the docs and it's not the proliferated `From` trait, I think that it's fine. I'll try to see if I can come up with another way that would only allow creating the sockets from within the crate.